### DISCUSSION TITLE
De centrale viewer en eigen node zijn niet 1:1

### DISCUSSION THREAD

**Discussion by CelineJansen** [kadaster-labs/sensrnet-home#305] [Q&A]
Node Eindhoven 68 sensoren:
![image](https://user-images.githubusercontent.com/79267472/160780833-d031e7c1-fa6b-4f2f-9900-fea2e9ceccfc.png)

Centrale viewer 67 sensoren:
![image](https://user-images.githubusercontent.com/79267472/160781156-a10f18db-ccad-47c1-adb3-86e00b5d2ab6.png)

(Ik zou 70 verwachten: 2 van Tilburg in Eindhoven + 68 van Eindhoven).

Waar kan dit aan liggen? Het lijkt geen caching te zijn, heb mijn browser geschiedenis verwijderd en het resultaat blijft hetzelfde. Via de API ($(Endpoint)device?legalEntityId=$(OrganisatieID)) krijg ik ze wel alle 68 door.

**Comment by marcvanandel** [comment]
Dat is gek ... Daar moeten we even induiken. Bedankt voor de melding, @CelineJansen . We kijken even wanneer we er naar kunnen kijken ...

**Reply by CelineJansen** [reply]
Bedankt! Mocht het helpen bij het onderzoeken van deze onvolkomenheid: ik heb gebruik gemaakt van de API zodat ik de sensoren in bulk op kon voeren.

**Reply by marcvanandel** [reply]
@CelineJansen Heb je voor mij een device ID waar ik naar kan kijken? Bijvoorbeeld waar ook de datastreams ontbreken lijkt mij wel interessant ...

**Reply by CelineJansen** [reply]
Datastreams die ontbreken in de centrale viewer terwijl ze bij mij wel gevuld zijn is bijvoorbeeld zo voor 8d11d974-5805-44db-ba0f-997f43a60928 (NO2 sensor 27).
1e39f6c0-f3e4-46b2-80ff-d87162cdf1d8 (NO2 sensor 9) is een voorbeeld van een device/sensor/datastroom combi die wel lokaal bij ons staat maar niet in de centrale viewer.
![image](https://user-images.githubusercontent.com/79267472/161260893-ef327640-2b7e-4e38-86f4-b01c3dd1026b.png)

**Reply by marcvanandel** [reply]
Die laatste kan ik wel terugvinden in onze eventstore. Dat betekent dat 'ie wel via de blockchain is doorgegeven en dus door de Eindhoven Registry Node is gepost én door onze Sync component in de Eventstore is gezet. Ik zie wel dat het eerste event niet het `DeviceRegistered` event is, maar `DeviceLocated`. Dat zou - volgens mij - niet mogen gebeuren ... (toch @kad-floriw ?)

![image](https://user-images.githubusercontent.com/2315396/161268675-41956f76-180f-4864-a37a-2bf77d97b5e4.png)

**Reply by marcvanandel** [reply]
Van die eerste zie ik maar drie events ... en dat zijn geen `DatastreamAdded` events ... Waar zijn die blijven hangen dan?

![image](https://user-images.githubusercontent.com/2315396/161269018-759ed02c-2bb0-4ea4-9178-d966c02c22bf.png)

Ik heb vanmorgen geprobeerd in de blockchain blokken te kijken ... maar dat is nog niet zo eenvoudig ...

**Reply by CelineJansen** [reply]
Als het helpt kan ik eventueel ook alles verwijderen en opnieuw opvoeren. Ik verwacht dat het dan is opgelost eerlijk gezegd... Heb ik puur niet gedaan omdat de bug misschien ook interessant kan zijn. Laat het maar weten wanneer het jullie handig lijkt als ik een nieuwe poging doe.

**Reply by marcvanandel** [reply]
In ieder geval zou het wél mogelijk moeten zijn dat elke projectie / view en dus ook de Centrale Viewer een andere volgorde van events ontvangen moet kunnen ondersteunen. Daarvoor heb ik bug #306 aangemaakt.

**Reply by marcvanandel** [reply]
Verwijderen betekent dat er nieuwe events worden toegevoegd, @CelineJansen. Het lijkt dan misschien wel 'opgelost' ... maar echt opgelost is het dan niet. En inderdaad heb je gelijk dat ik vind dat we moeten weten waarom het niet goed gaat ... Dat blijkt helaas wat hardnekkiger dan ik zou willen 😏 😁

**Reply by marcvanandel** [reply]
Van `8d11d974-5805-44db-ba0f-997f43a60928` zijn de events voor `DatasteamAdded` niet te vinden in de blockchain. Blijkbaar zijn deze (dus?) niet goed ge'sync'd in de Eindhoven Registry Node ... 🤔 Interessant ... en moeilijk om daar nu achter te komen, ben ik bang.

**Reply by CelineJansen** [reply]
Zou het kunnen dat er door de bulk opvoer teveel events in een keer kort achterelkaar worden afgevuurd waardoor de synch het moeilijk heeft? Dat het dus een prestatie dingetje is zeg maar?

**Reply by marcvanandel** [reply]
Hmm... dat is niet onze ervaring, zie https://github.com/kadaster-labs/sensrnet-home/issues/189. Daar hebben we zelf geautomatiseerd tot wel 300.000 sensoren opgevoerd ... zonder echte issues 💪

**Reply by CelineJansen** [reply]
> Hmm... dat is niet onze ervaring, zie #189. Daar hebben we zelf geautomatiseerd tot wel 300.000 sensoren opgevoerd ... zonder echte issues 💪

O wow, dat is tof zeg!

**Reply by marcvanandel** [reply]
Just for the record: Er zijn alleen maar v1 events in de chain opgenomen ... en waarschijnlijk zijn er ook alleen maar Registry Nodes verbonden die v1 events produceren en lezen. Daar kan het probleem dus niet in zitten ...

**Reply by marcvanandel** [reply]
Om enige checks uit te voeren (zie https://github.com/kadaster-labs/sensrnet-home/discussions/305#discussioncomment-2486051), heb ik contact gezocht met Conduction ...

**Reply by marcvanandel** [reply]
Events in de Eventstore van Eindhoven Registry Node:

![image](https://user-images.githubusercontent.com/2315396/161285674-93d012f4-f606-4903-bf25-90ea7fb0fddf.png)

**Reply by marcvanandel** [reply]
Blijkbaar is de `DatastreamAdded` niet gepubliceerd op de blockchain ... op een of andere reden 🤔

**Reply by marcvanandel** [reply]
Er lijken twee problemen te spelen:

1. In het cluster zijn de netwerkverbindingen niet helemaal stabiel. Pods lijken vaker opnieuw op te starten dan is te verwachten -> verder onderzoek door Conduction & partners is hier nodig
2. De Sync bridge 'mist' op de een of andere manier - waarschijnlijk / mogelijk door de netwerk issues - events in het kopiëren van de Eventstore naar de blockchain

Een mogelijk onderliggend probleem kan zijn dat de Eventstore pods - aantal parallelle containers die er draaien - niet helemaal stabiel zijn. Er lijken 'leadership elections' plaats te vinden ... die we niet verwachten, hoogstens bij opstarten van het systeem.

**Reply by marcvanandel** [reply]
De Eventstore is niet ons eigen component ... en daar willen we geen synchronisatie issues van oplossen ... bij voorkeur. Oplossing - voor korte termijn - zou kunnen zijn om de Evenstore als enkele instantie te draaien in het cluster. Dat gaat ten koste van de beschikbaarheid en robuustheid van het systeem ... alhoewel er nu al vreemde issues optreden bij 3 instanties. Beter om de Eventstore minder beschikbaar te maken en daarmee minder fragiel vanuit schaling ('leader elections' etc) en risico's op algemene down time.

Mocht de Eventstore helemaal in de soep lopen, zou vanuit de Multichain alle events geladen moeten kunnen worden. Dit ondersteunt de SyncBridge nu nog niet, maar zou toegevoegd kunnen worden in de `abstract.mc.consumer.ts` (line 45 zou een `else` kunnen hebben op basis van een nieuwe 'flag' `recoveryMode` oid)

**Comment by CelineJansen** [comment]
Aanvullend: de datastromen komen ook niet goed mee. 

Dit is onze registratie node:
![image](https://user-images.githubusercontent.com/79267472/160815391-ae644d75-86de-4300-92e7-a4594e8e017a.png)

Dit is de centrale viewer:
![image](https://user-images.githubusercontent.com/79267472/160815592-e232cef5-a0ca-4bc7-bee3-003dfae77f23.png)
